plasmapy/utils/tests/test_code_repr.py

```diff
@@ -0,0 +1,370 @@
+"""Tests for code representation functions."""
+
+import numpy as np
+import pytest
+
+from astropy import units as u
+
+from plasmapy.utils.code_repr import (
+    _code_repr_of_arg,
+    _code_repr_of_args_and_kwargs,
+    _code_repr_of_ndarray,
+    _code_repr_of_quantity,
+    _name_with_article,
+    _object_name,
+    _string_together_warnings_for_printing,
+    attribute_call_string,
+    call_string,
+    method_call_string,
+)
+
+
+def generic_function(*args, **kwargs):
+    return None
+
+
+def adams_number():
+    return 42
+
+
+# function, args, kwargs, expected
+call_string_table = [
+    (generic_function, (), {}, "generic_function()"),
+    (generic_function, (1), {}, "generic_function(1)"),
+    (generic_function, ("x"), {}, "generic_function('x')"),
+    (generic_function, (1, "b", {}), {}, "generic_function(1, 'b', {})"),
+    (generic_function, (), {"kw": 1}, "generic_function(kw=1)"),
+    (generic_function, (), {"x": "c"}, "generic_function(x='c')"),
+    (
+        generic_function,
+        (1, 2, 3),
+        {"a": 4, "b": 5, "c": 6},
+        "generic_function(1, 2, 3, a=4, b=5, c=6)",
+    ),
+    (
+        generic_function,
+        np.array((1.0, 2.0)),
+        {},
+        "generic_function(np.array([1., 2.]))",
+    ),
+    (
+        generic_function,
+        np.array((3.0, 4.0)) * u.m,
+        {},
+        "generic_function(np.array([3., 4.])*u.m)",
+    ),
+    (
+        generic_function,
+        (ValueError,),
+        {"x": TypeError},
+        "generic_function(ValueError, x=TypeError)",
+    ),
+    (generic_function, (), {"bowl": super}, "generic_function(bowl=super)"),
+    (
+        generic_function,
+        (1, "b"),
+        {"b": 42, "R2": "D2"},
+        "generic_function(1, 'b', b=42, R2='D2')",
+    ),
+]
+
+
+@pytest.mark.parametrize("function,args,kwargs,expected", call_string_table)
+def test_call_string(function, args, kwargs, expected):
```

Comment:
Unless `call_string_table` (and the like below) is (are) going to be called in several locations, I find that this disassociation of the parametrized list and the test function to significantly hinder readability of tests.  I mean, I start reading the test and then I have to wonder away from the test to get the cases that are being tested.

Reply:
I think the main problem with readability here is that the when we parametrize tests using a list containing a series of tuples which might contain inner tuples, the parameters get detached from the corresponding semantic information/metadata.  It'd be so much more readable if the arguments were defined like keywords.  In #925 I define classes that are intended to address this problem...without me having to rewrite essentially the same code over and over again.  I might try doing something like that here, though independent of #925, since I'm trying to avoid using the test helpers to test the test helpers.

Reply:
I agree with that sentiment.  `pytest.mark.parametrize`'s separation of names and values does make it a bit difficult to follow.

My point here is that when someone reads `@pytest.mark.parametrize("function,args,kwargs,expected", call_string_table)` they have to immediately search the code to find `call_string_table`.  The fact that I have to leave the test to find out what is being tested is what reduces readability for me.